### PR TITLE
feat(config): add config handler

### LINKED ISSUE #4: Write config parser
The config file should be in the [toml](https://toml.io/en/) format, and should be stored in `/etc/pacstall/config.toml` (allow this location to be customized).

Progress:
* https://github.com/pacstall/libpacstall/pull/13

### LINKED ISSUE #5: Setup ReadTheDocs

### LINKED ISSUE #17: Implement a config validation function
Validates the following data in the config:
* Valid toml
* Data types
* Correct placement of keys (so no `jobs` key in `repository` table for example)
* The values (validates `url`s etc)

### LINKED ISSUE #19: Implement `sources` table in the config parser
Here's the last agreed upon template of the table:
```toml
[sources.official]
url = "https://github.com/pacstall/pacstall-programs"
aliases = ["main", "etc"]
preference = 1

[sources.custom1]
url = "customUrl1"
aliases = ["hydra", "custom alias 2"]
preference = 2
```

Here's how the raw parsed dictionary looks like:

```python
{
│   'sources': {
│   │   'official': {'url': 'https://github.com/pacstall/pacstall-programs', 'aliases': ['main', 'etc'], 'preference': 1},
│   │   'custom1': {'url': 'customUrl1', 'aliases': ['hydra', 'custom alias 2'], 'preference': 2}
│   }
}
```

### LINKED ISSUE #25: Setup testing with tox

### PR DESCRIPTION
**Pull request by wizard-28**
This is the first feature that's ported to Rust.

Re-implements #11 
Re-implements #12 
Re-implements #6

Closes #5 
Closes #25 
Closes #19
Closes #17 
Closes #4

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+    fn default() -> Self {
+        Self {
+            name: "official".into(),
+            url: "https://github.com/pacstall/pacstall-programs".into(),
+            preference: 1,
+        }
+    }
+}
+
+#[cfg(test)]
```
Comment (Elsie19) [inline src/config.rs:227]: Do we include tests in the config.rs or should we make a separate file for testing?

Comment (wizard-28) [reply src/config.rs:227]: All unit tests for the config handler is written in the `tests` module in the `config.rs` file. The unit tests are in the same file as the code.

File: src/config.rs [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+#[derive(Deserialize, Debug, Eq, PartialEq, Serialize)]
+#[serde(default, deny_unknown_fields)]
+pub struct Settings {
+    /// The preferred `editor` set by the user. (Defaults to the environment
+    /// variables *EDITOR* -> *VISUAL* -> the editor chosen by the
+    /// `sensible-editor` command.)
+    pub editor: String,
+    /// Number of threads to use for building packages (Defaults to the number
+    /// of *logical cores* on the computer.)
+    pub jobs: u8,
```
Comment (volitank) [inline src/config.rs:165]: What happens with `serde` if someone specifies the editor string but not the jobs?

Does it error, set the default to 0 or use the number of CPU?

Comment (wizard-28) [reply src/config.rs:165]: It defaults to the number of logical cores on the computer, as documented.

File: src/config.rs [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+}
+
+/// The extracted configuration.
+#[derive(Deserialize, Debug, Default, Eq, PartialEq, Serialize)]
+#[serde(deny_unknown_fields)]
+pub struct Config {
+    #[serde(default)]
+    pub settings: Settings,
+    #[serde(default = "default_repository")]
+    pub repositories: Vec<Repository>,
```
Comment (saenai255) [inline src/config.rs:107]: default repository is a list?

File: src/config.rs [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+    ///
+    /// Note that the URL **isn't verified** during extraction!
+    pub url: String,
+    /// Preference of the repository.
+    ///
+    /// Specifies which repository to look into first during certain operations
+    /// like installing a package. If the package isn't present in the first
+    /// preferred repository, then the second preferred repository is looked
+    /// into.
+    pub preference: u32,
```
Comment (saenai255) [inline src/config.rs:212]: maybe `u8`?

**Review by volitank — APPROVED** [review]
I approve of these nice changes!

**Review by D-Brox — APPROVED** [review]
Wizard never fails to impress

**Review by saenai255 — APPROVED** [review]

File: src/config.rs (hunk 4 above)
Comment (wizard-28) [reply src/config.rs:212]: I mean we can but why?

Comment (volitank) [reply src/config.rs:212]: I personally try to stick with the one that my data is going to fit in most closely. So If I only need to work with say `1, 2, 3`, then I would use u8.

In terms of performance, I don't actually think it matters. Maybe uses a slight bit more memory to allocate more bits on the stack for the number, but would be insignificant in the grand scheme of things.

`u8` is most used in rust to indicate bytes, like a byte string is `Vec<u8>`.

**Pull request merged**